reference/test/solver/bicgstab_kernels.cpp

```diff
@@ -67,168 +69,205 @@ class Bicgstab : public ::testing::Test {
 }
 
 
-TEST_F(Bicgstab, SolvesMultipleDenseSystemsUsingAdvancedApply)
+TYPED_TEST(Bicgstab, SolvesMultipleDenseSystemsUsingAdvancedApply)
 {
-    auto solver = bicgstab_factory->generate(mtx);
-    auto alpha = gko::initialize<Mtx>({2.0}, exec);
-    auto beta = gko::initialize<Mtx>({-1.0}, exec);
-    auto b =
-        gko::initialize<Mtx>({{-1.0, -5.0}, {3.0, 1.0}, {1.0, -2.0}}, exec);
-    auto x = gko::initialize<Mtx>({{0.5, 1.0}, {1.0, 2.0}, {2.0, 3.0}}, exec);
+    using Mtx = typename TestFixture::Mtx;
+    using value_type = typename TestFixture::value_type;
+    using T = value_type;
+    auto half_tol = std::sqrt(r<value_type>::value);
+    auto solver = this->bicgstab_factory->generate(this->mtx);
+    auto alpha = gko::initialize<Mtx>({2.0}, this->exec);
+    auto beta = gko::initialize<Mtx>({-1.0}, this->exec);
+    auto b = gko::initialize<Mtx>(
+        {I<T>{-1.0, -5.0}, I<T>{3.0, 1.0}, I<T>{1.0, -2.0}}, this->exec);
+    auto x = gko::initialize<Mtx>(
+        {I<T>{0.5, 1.0}, I<T>{1.0, 2.0}, I<T>{2.0, 3.0}}, this->exec);
 
     solver->apply(alpha.get(), b.get(), beta.get(), x.get());
 
 
-    GKO_ASSERT_MTX_NEAR(x, l({{-8.5, 1.0}, {-3.0, 2.0}, {6.0, -5.0}}), 1e-8);
+    GKO_ASSERT_MTX_NEAR(x, l({{-8.5, 1.0}, {-3.0, 2.0}, {6.0, -5.0}}),
+                        half_tol);
 }
 
 
 // The following test-data was generated and validated with MATLAB
-TEST_F(Bicgstab, SolvesBigDenseSystemForDivergenceCheck1)
+TYPED_TEST(Bicgstab, SolvesBigDenseSystemForDivergenceCheck1)
 {
+    using Mtx = typename TestFixture::Mtx;
+    using value_type = typename TestFixture::value_type;
+    auto half_tol = std::sqrt(r<value_type>::value);
     std::shared_ptr<Mtx> locmtx =
         gko::initialize<Mtx>({{-19.0, 47.0, -41.0, 35.0, -21.0, 71.0},
                               {-8.0, -66.0, 29.0, -96.0, -95.0, -14.0},
                               {-93.0, -58.0, -9.0, -87.0, 15.0, 35.0},
                               {60.0, -86.0, 54.0, -40.0, -93.0, 56.0},
                               {53.0, 94.0, -54.0, 86.0, -61.0, 4.0},
                               {-42.0, 57.0, 32.0, 89.0, 89.0, -39.0}},
-                             exec);
-    auto solver = bicgstab_factory_precision->generate(locmtx);
-    auto b = gko::initialize<Mtx>({0.0, -9.0, -2.0, 8.0, -5.0, -6.0}, exec);
-    auto x = gko::initialize<Mtx>({0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, exec);
+                             this->exec);
+    auto solver = this->bicgstab_factory_precision->generate(locmtx);
+    auto b =
+        gko::initialize<Mtx>({0.0, -9.0, -2.0, 8.0, -5.0, -6.0}, this->exec);
+    auto x = gko::initialize<Mtx>({0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, this->exec);
 
     solver->apply(b.get(), x.get());
 
     GKO_ASSERT_MTX_NEAR(
         x,
         l({0.13853406350816114, -0.08147485210505287, -0.0450299311807042,
            -0.0051264177562865719, 0.11609654300797841, 0.1018688746740561}),
-        1e-9);
+        half_tol * 5e-1);
```

Comment:
Is any reason to times 0.5 here?

Reply:
These are pretty much always the same sort of issues, you need minor adjustments to the precision to make it work in all cases, including single.